pkg/controller/nodesetup/controller.go

```diff
@@ -0,0 +1,238 @@
+
+	nodeConfigLister scyllav1alpha1listers.NodeConfigLister
+
+	cachesToSync []cache.InformerSynced
+
+	eventRecorder record.EventRecorder
+
+	queue    workqueue.RateLimitingInterface
+	handlers *controllerhelpers.Handlers[*scyllav1alpha1.NodeConfig]
+
+	nodeName       string
+	nodeUID        types.UID
+	nodeConfigName string
+	nodeConfigUID  types.UID
+
+	executor           exec.Interface
+	systemdControl     *systemd.SystemdControl
+	systemdUnitManager *systemd.UnitManager
+	sysfsPath          string
+}
+
+func NewController(
+	ctx context.Context,
+	kubeClient kubernetes.Interface,
+	scyllaClient scyllav1alpha1client.ScyllaV1alpha1Interface,
+	nodeConfigInformer scyllav1alpha1informers.NodeConfigInformer,
+	nodeName string,
+	nodeUID types.UID,
+	nodeConfigName string,
+	nodeConfigUID types.UID,
+) (*Controller, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartStructuredLogging(0)
+	eventBroadcaster.StartRecordingToSink(&corev1client.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+
+	if kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage(
+			"nodesetup_controller",
+			kubeClient.CoreV1().RESTClient().GetRateLimiter(),
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	systemdControl, err := systemd.NewSystemdSystemControl(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("can't create systemd control: %w", err)
+	}
+
+	ncc := &Controller{
+		kubeClient:   kubeClient,
+		scyllaClient: scyllaClient,
+
+		nodeConfigLister: nodeConfigInformer.Lister(),
+
+		cachesToSync: []cache.InformerSynced{
+			nodeConfigInformer.Informer().HasSynced,
+		},
+
+		eventRecorder: eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "nodesetup-controller"}),
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "NodeSetup"),
+
+		nodeName:       nodeName,
+		nodeUID:        nodeUID,
+		nodeConfigName: nodeConfigName,
+		nodeConfigUID:  nodeConfigUID,
+
+		executor:           exec.New(),
+		systemdControl:     systemdControl,
+		systemdUnitManager: systemd.NewUnitManager("scylla-operator-node-setup"),
+		sysfsPath:          "/sys",
+	}
+
+	ncc.handlers, err = controllerhelpers.NewHandlers[*scyllav1alpha1.NodeConfig](
+		ncc.queue,
+		keyFunc,
+		scheme.Scheme,
+		nodeConfigControllerGVK,
+		kubeinterfaces.GlobalGetList[*scyllav1alpha1.NodeConfig]{
+			GetFunc: func(name string) (*scyllav1alpha1.NodeConfig, error) {
+				return ncc.nodeConfigLister.Get(name)
+			},
+			ListFunc: func(selector labels.Selector) (ret []*scyllav1alpha1.NodeConfig, err error) {
+				return ncc.nodeConfigLister.List(selector)
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("can't create handlers: %w", err)
+	}
+
+	nodeConfigInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ncc.addNodeConfig,
+		UpdateFunc: ncc.updateNodeConfig,
+		DeleteFunc: ncc.deleteNodeConfig,
+	})
+
```

Comment:
what if a new Node is added?

Reply:
These controllers are deployed per Node, Node changes aren't interesting here